src/pybind/mgr/insights/module.py

```diff
@@ -0,0 +1,254 @@
+
+        self.log.info("Maybe flushing slot {} needed {}".format(
+            self._health_slot, self._health_slot.need_flush()))
+
+        if self._health_slot.need_flush():
+            key = self._health_slot.key()
+
+            # build store data entry
+            slot = self._health_slot.health()
+            assert "version" not in slot
+            slot.update(dict(version = ON_DISK_VERSION))
+            data = json.dumps(slot, cls=HealthEncoder)
+
+            self.log.debug("Storing health key {} data {}".format(
+                key, json.dumps(slot, indent=2, cls=HealthEncoder)))
+
+            self.set_store(key, data)
+            self._health_slot.mark_flushed()
+
+    def _health_filter(self, f):
+        """Filter hourly health reports timestamp"""
+        matches = filter(
+            lambda t: f(HealthHistorySlot.key_to_time(t[0])),
+            six.iteritems(self.get_store_prefix(HEALTH_HISTORY_KEY_PREFIX)))
+        return map(lambda t: t[0], matches)
+
+    def _health_prune_history(self):
+        """Prune old health entries"""
+        cutoff = datetime.datetime.utcnow() - \
+                datetime.timedelta(hours = HEALTH_RETENTION_HOURS)
+        for key, _ in self._health_filter(lambda ts: ts <= cutoff):
+            self.log.info("Removing old health slot key {}".format(key))
+            self.set_store(key, None)
+
+    def _health_report(self, hours):
+        """
+        Report a consolidated health report for the past N hours.
+        """
+        # roll up the past N hours of health info
+        collector = HealthHistorySlot()
+        keys = HealthHistorySlot.key_range(hours)
+        for key in keys:
+            data = self.get_store(key)
+            self.log.info("Reporting health key {} found {}".format(
+                key, bool(data)))
+            health = json.loads(data) if data else {}
+            slot = HealthHistorySlot(health)
+            collector.merge(slot)
+
+        # fold in the current health
+        curr_health = json.loads(self.get("health")["json"])
+        curr_health_slot = HealthHistorySlot()
+        curr_health_slot.add(curr_health)
+        collector.merge(curr_health_slot)
+
+        return dict(
+           current = curr_health,
+           history = collector.health()
+        )
+
+    def _version_parse(self, version):
+        """
+        Return the components of a Ceph version string.
+
+        This returns nothing when the verison string cannot be parsed into its
+        constituent components, such as when Ceph has been built with
+        ENABLE_GIT_VERSION=OFF.
+        """
+        r = "ceph version (?P<release>\d+)\.(?P<major>\d+)\.(?P<minor>\d+)"
+        m = re.match(r, version)
+        ver = {} if not m else {
+            "release": m.group("release"),
+            "major": m.group("major"),
+            "minor": m.group("minor")
+        }
+        return { k:int(v) for k,v in ver.iteritems() }
```

Comment:
might want to use `six.iteritems()`?

Reply:
thanks. got that one, and caught another.